go_covert_lib/controller/processor/symmetricEncryption/symmetricEncryption.go

```
@@ -0,0 +1,237 @@
+
+	stream := cipher.NewCFBEncrypter(block, iv)
+	stream.XORKeyStream(cipherText[blockSize:], data)
+
+	return cipherText[:]
+}
+
+func CTREncrypter(block cipher.Block, data []byte, blockSize int) []byte {
+	cipherText := make([]byte, blockSize+len(data))
+	iv := cipherText[:blockSize]
+
+	stream := cipher.NewCTR(block, iv)
+	stream.XORKeyStream(cipherText[blockSize:], data)
+
+	return cipherText[:]
+}
+
+func Pad(data []byte) []byte {
+	// padding is done by having 7 zero's at the start followed by the length 
+	// of the original message, the contents of the message, followed by a 
+	// series of zero's to fill the block size requirements of the symmetric  
+	// encryption algorithm 
+	lenOfData := len(data)
+	data = append(data, 0)
+	copy(data[1:], data)
+	data[0] = byte(lenOfData)
+
+	for i := 0; i < padIndex; i++ {
+		data = append(data, 0)
+		copy(data[1:], data)
+		data[0] = 0
+	}
+
+	for len(data)%aes.BlockSize != 0 {
+		data = append(data, 0)
+	}
+
+	return data[:]
+}
+
+func UnPad(data []byte) []byte {
+	lenOfData := data[padIndex]
+	data = data[padIndex+1:]
+	return data[:lenOfData]
+}
+
+func (c *SymmetricEncryption) Unprocess(data []byte) ([]byte, error) {
+	algorithm := c.algorithm
+
+	var (
+		blockSize int
+		block     cipher.Block
+		err       error
+	)
+
+	// based on the users choice of symmetric algorithm create a cipher
+	switch algorithm {
+	case "Advanced Encryption Standard (AES)":
+		block, err = aes.NewCipher(c.key)
+		blockSize = 16
+	case "Data Encryption Standard (DES)":
+		block, err = des.NewCipher(c.key)
+		blockSize = 8
+	case "Triple Data Encryption Standard (3DES)":
+		block, err = des.NewTripleDESCipher(c.key)
+		blockSize = 8
+	default:
+		return nil, errors.New("Undefined algorithm selected")
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	// based on the users choice of the mode of operation encrypt in that mode
+	selectedMode := c.mode
+	switch selectedMode {
+	case "Galois Counter Mode (GCM)":
```

As mentioned above, the GCMDecryptor scenario should return the error it encounters and return that error message.